src/core/jit/rules/query_depth.rs

```diff
@@ -0,0 +1,118 @@
+use super::ExecutionRule;
+use crate::core::jit::{Field, Nested, OperationPlan};
+use crate::core::valid::Valid;
+
+pub struct QueryDepth(usize);
+
+impl QueryDepth {
+    pub fn new(depth: usize) -> Self {
+        Self(depth)
+    }
+}
+
+impl ExecutionRule for QueryDepth {
```

Comment:
ExecutionRule is basically Transform with Value = OperationPlan and Error = String. Let's reuse it here.